### DISCUSSION TITLE
Making Broadcasting More Explicit

### DISCUSSION THREAD

**Discussion by Pratham-ja** [harvard-edge/cs249r_book#1187] [TinyTorch Q&A]
Hi Professor VJ and everyone else. While going through the loss module (module 04) I came across something which I used to struggle when I was learning torch using numpy from scratch. Many times there is a mathematical mistake in our operations and we are trying some element wise operation but the **hidden broadcasting flexibility does not raise any error and what seems to be correct without any error raised by the numpy or torch, can lead to wrong results** and not always what a student is trying to do with the tensors. 
So, for example: 
If I have some final prediction data and we are trying to find the mean square loss on the 32 examples and 4 output values (for example in multiple linear regression problems) then if we have 
pred.shape = (32,4) and y_true.shape(4) due to some mistakes or typos I myself many times miss some dimensions while aggregating along some axis and in general) then in the implementation MSELoss class we calculate   
diff = predictions.data - targets.data 
This allows shape mismatches like:
pred: (32, 4)
target: (4,) 
which will silently broadcast into (32,4) and we end up using the same y_true for all the examples. Although **the scikit api for the mean_square_error will raise the shape mismatch error**.
While TinyTorch mirrors PyTorch behavior, but sine it is more for the education purposes, exposing clearly what is going under the hood for every line of code, I believe the explicit shape matching should be enforced to prevent unintended mistakes due to broadcasting. This makes the individual understand the broadcasting in a better way and will use this feature more consciously.
Also **it can be made very general in TinyTorch that whenever broadcasting occurs in some operation we can print a message that this operation involves broadcasting along some axis and this can make the user more informed and careful**.
Would love to know @profvjreddi and others thought on this.

**Comment by profvjreddi** [comment]
Really thoughtful suggestion @Pratham-ja, and I've been bitten by silent broadcasting bugs myself.

You're right that this is a real problem. We've added a fix in Module 01 (Tensor) where broadcasting is first taught. Students now see both how broadcasting works *and* how it can silently produce wrong results when shapes are accidentally mismatched. The arithmetic unit test now includes a shape-safety check that demonstrates the exact scenario you described — predictions `(32, 4)` vs targets `(4,)` broadcasting silently.

The change is on the `fix/loss-shape-warnings` branch and teaches students to always verify `predictions.shape == targets.shape` before computing loss. We put it in Module 01 rather than Module 04 because broadcasting is a tensor-level concept — by the time students hit loss functions, they should already have the habit of checking shapes.

Thanks for raising this. Would you be interested in contributing further? For example, an optional verbose mode that prints when broadcasting occurs could be a great addition to the Tensor class itself.

**Comment by Pratham-ja** [comment]
I’d be happy to contribute further @profvjreddi. I’ll open an issue with the ENH label to propose this improvement and work on extending the Tensor class. We can continue any detailed discussion there.

**Comment by profvjreddi** [comment]
Hi @Pratham-ja 

I actually addressed this in Module 01 (Tensor) earlier today, which is where students first encounter broadcasting. I put it there deliberately because the best time to learn about a footgun is the moment you're handed the tool. So Module 01 now includes a "Broadcasting Pitfall" warning showing exactly the (32, 4) - (4,) scenario you described, plus a shape-check assertion in the unit test demonstrating the defensive pattern. Check it out and let me know if you think its good. It is https://github.com/harvard-edge/cs249r_book/commit/04f8d0efa399956430b8d313955f3e92cab66f43 and if you are good with it we can close this thread, and if not we can discuss further.

**Comment by profvjreddi** [comment]
Marking this as answered since the shape-safety checks and broadcasting warnings landed in 04f8d0e. @Pratham-ja feel free to open a separate enhancement issue if you'd like to extend the Tensor class further. Thanks again for finding all these nitty gritty issues 🙏
